util/chunk/chunk.go

```diff
@@ -112,12 +151,73 @@ func (c *Chunk) SetNumVirtualRows(numVirtualRows int) {
 // Reset resets the chunk, so the memory it allocated can be reused.
 // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
 func (c *Chunk) Reset() {
-	for _, c := range c.columns {
-		c.reset()
+	c.GrowReset(c.capacity)
+}
+
+// GrowReset resets the chunk and grow capacity if need.
+// TODO: this method will be used in continue PR.
+func (c *Chunk) GrowReset(capLimit int) {
+	if c.columns == nil {
+		return
+	}
+	needResize, nextCapacity := checkCapacity(c, capLimit, false)
+	if needResize {
+		c.capacity = nextCapacity
+	}
+	for _, col := range c.columns {
+		if needResize {
+			col.eraseGrow(nextCapacity)
+		} else {
+			col.reset()
+		}
 	}
 	c.numVirtualRows = 0
 }
 
+// checkCapacity checks and return suitable capacity in next execution.
+func checkCapacity(c *Chunk, capLimit int, tryReduce bool) (needResize bool, nextCapacity int) {
+	currSize, nextCapacity := c.NumRows(), c.capacity
+	if currSize == 0 {
+		return false, nextCapacity
+	}
+	if currSize >= c.capacity && currSize < capLimit {
+		nextCapacity = c.capacity * 2
+		if nextCapacity > capLimit {
+			nextCapacity = capLimit
+		}
+		return true, nextCapacity
+	}
+	if tryReduce && currSize < c.capacity {
+		nextCapacity = powerOf2Capacity(currSize, capLimit)
+		if nextCapacity > capLimit {
+			nextCapacity = capLimit
+		}
+	}
+	return false, nextCapacity
+}
+
+// Returns a power of two size for the given target size.
+func powerOf2Capacity(size, capLimit int) int {
+	n := size - 1
```

Comment:
I think we'd better extract 
line 201~206 as to a util function like NextPowerOfTwo